### DISCUSSION TITLE
How to handle error from middleware

### DISCUSSION THREAD

**Discussion by SardineFish** [actix/actix-web#2750] [Q&A]
## Problem
In the middleware service chain, we call next service with `ServiceRequest` and return the `ServiceResponse` from the next service.
But if next service return an error, how can we make it into a `ServiceResponse`

## Background
In our project, we need to make our APIs always response in a certain JSON format. 
But if we have error in middleware, we can only return it out and let actix-web response an `Internal Serivce Error`.
We cannot make a custom `ServiceResponse` because `ServiceRequest` is moved to next service and the inner `HttpRequest` cannot clone.

For example in official middleware `ErrorHandler`  line 157
https://github.com/actix/actix-web/blob/6a5b37020676fdfed4b8c7466d8542904bca825c/actix-web/src/middleware/err_handlers.rs#L148-L177
The error from next service is just returned to upper service.
Finally it will become a `500` response with default text body `Internal Serivce Error`.

**Comment by fakeshadow** [comment]
In general you don't.
Most of middlewares would never use error path and it's just a placeholder. You can handle error if you write multiple middlewares and nest them but don't expect other middlewares would follow the suite.(In that case you use Error(HttpRequest) and unlike you stated it's Clone https://docs.rs/actix-web/4.0.1/actix_web/struct.HttpRequest.html#impl-Clone )

**Reply by SardineFish** [reply]
But the `Scope::wrap<M, B>` restrict the middleware to implement trait 
`Transform<
                T::Service,
                ServiceRequest,
                Response = ServiceResponse<B>,
                Error = Error,
                InitError = (),
            > + 'static`
The associated type `Error` is constraint to `actix_web::Error` so I cannot use my custom error type that wraps the `HttpRequest`

**Reply by fakeshadow** [reply]
https://docs.rs/actix-web/latest/actix_web/struct.Error.html#method.as_error.
`actix_web::Error` is a trait object for `ResponseError`. You can downcast the object to the real type impl `ResponseError`.
